sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -131,181 +164,160 @@ public boolean isEndRequired() {
-              // the batch. It is acceptable because batching is a best effort mechanism here.
-              try {
-                monitor.wait(scheduleDelayMillis);
-              } catch (InterruptedException ie) {
-                // Preserve the interruption status as per guidance and stop doing any work.
-                Thread.currentThread().interrupt();
-                return;
-              }
-            } while (spansList.isEmpty());
+    final CompletableResultCode result = new CompletableResultCode();
+    executor.submit(
+        new Runnable() {
+          @Override
+          public void run() {
+            exportNextBatch(spanBuffer, 0, result);
           }
-          spansCopy = new ArrayList<>(spansList);
-          spansList.clear();
-        }
-        // Execute the batch export outside the synchronized to not block all producers.
-        exportBatches(spansCopy);
-      }
-    }
-
-    private void shutdown() {
-      forceFlush();
-      timer.cancel();
-      spanExporter.shutdown();
-    }
+        });
+
+    result.whenComplete(
+        new Runnable() {
+          @Override
+          public void run() {
+            if (queue.size() >= maxExportBatchSize) {
+              executor.submit(
+                  new Runnable() {
+                    @Override
+                    public void run() {
+                      exportSpans();
+                    }
+                  });
+            } else {
+              scheduleNextExport();
+            }
+          }
+        });
+  }
 
-    private void forceFlush() {
-      ArrayList<ReadableSpan> spansCopy;
-      synchronized (monitor) {
-        spansCopy = new ArrayList<>(spansList);
-        spansList.clear();
-      }
-      // Execute the batch export outside the synchronized to not block all producers.
-      exportBatches(spansCopy);
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void exportNextBatch(
+      final List<ReadableSpan> spans, final int offset, final CompletableResultCode result) {
+    int num = Math.min(maxExportBatchSize, spans.size() - offset);
+    final int limit = offset + num;
+    List<SpanData> forExport = new ArrayList<>(num);
+    for (int i = offset; i < limit; i++) {
+      forExport.add(spans.get(i).toSpanData());
     }
-
-    private void exportBatches(ArrayList<ReadableSpan> spanList) {
-      // TODO: Record a counter for pushed spans.
-      for (int i = 0; i < spanList.size(); ) {
-        int lastIndexToTake = Math.min(i + maxExportBatchSize, spanList.size());
-        onBatchExport(createSpanDataForExport(spanList, i, lastIndexToTake));
-        i = lastIndexToTake;
-      }
+    final CompletableResultCode exportResult;
+    try {
+      exportResult = spanExporter.export(forExport);
+    } catch (Exception e) {
+      logger.log(Level.FINE, "Exporter failed", e);
+      maybeContinueBatch(spans, limit, result);
+      return;
     }
+    exportResult.whenComplete(
```

Comment:
What do you think about putting this inside the try block? Seems like it might read a little more cleanly not to have to pre-declare the `exportResult`.